cpp/src/parquet/arrow/path_internal.cc

```diff
@@ -0,0 +1,759 @@
+    }
+    if (ARROW_PREDICT_FALSE(result == kError)) {
+      DCHECK(!context.last_status.ok());
+      return context.last_status;
+    }
+    stack_position += static_cast<int>(result);
+  }
+  RETURN_NOT_OK(context.last_status);
+  builder_result.def_rep_level_count = context.def_levels.length();
+
+  if (context.rep_levels.length() > 0) {
+    builder_result.rep_levels = context.rep_levels.data();
+    std::swap(builder_result.post_list_visited_elements, context.visited_elements);
+    if (builder_result.post_list_visited_elements.empty()) {
+      builder_result.post_list_visited_elements.push_back({0, 0});
+    }
+  } else {
+    builder_result.post_list_visited_elements.push_back(
+        {0, builder_result.leaf_array->length()});
+    builder_result.rep_levels = nullptr;
+  }
+
+  builder_result.def_levels = context.def_levels.data();
+  return writer(builder_result);
+}
+
+struct FixupVisitor {
+  int max_rep_level = -1;
+  std::vector<PathInfo::Node> cleaned_up_nodes;
+  int16_t rep_level_if_null = kLevelNotSet;
+
+  template <typename T>
+  void HandleListNode(T& arg) {
+    if (arg.rep_level() == max_rep_level) {
+      arg.SetLast();
+      // after the last list node we don't need to fill
+      // rep levels on null.
+      rep_level_if_null = kLevelNotSet;
+    } else {
+      rep_level_if_null = arg.rep_level();
+    }
+  }
+  void operator()(ListNode& node) { HandleListNode(node); }
+  void operator()(LargeListNode& node) { HandleListNode(node); }
+  void operator()(FixedSizeListNode& node) { HandleListNode(node); }
+
+  // For non-list intermediate nodes.
+  template <typename T>
+  void HandleIntermediateNode(T& arg) {
+    if (rep_level_if_null != kLevelNotSet) {
+      arg.SetRepLevelIfNull(rep_level_if_null);
+    }
+  }
+
+  void operator()(NullableNode& arg) { HandleIntermediateNode(arg); }
+
+  void operator()(AllNullsTerminalNode& arg) {
+    // Even though no processing happens past this point we
+    // still need to adjust it if a list occurred after an
+    // all null array.
+    HandleIntermediateNode(arg);
+  }
+
+  void operator()(NullableTerminalNode& arg) {}
+  void operator()(AllPresentTerminalNode& arg) {}
+};
+
+PathInfo Fixup(PathInfo info) {
+  // We only need to fixup the path if there were repeated
+  // elements on it.
+  if (info.max_rep_level == 0) {
+    return info;
+  }
+  FixupVisitor visitor;
+  visitor.max_rep_level = info.max_rep_level;
+  if (visitor.max_rep_level > 0) {
+    visitor.rep_level_if_null = 0;
+  }
+  visitor.cleaned_up_nodes.reserve(info.path.size());
+  for (size_t x = 0; x < info.path.size(); x++) {
+    ::arrow::util::visit(visitor, info.path[x]);
+  }
+  return info;
+}
+
+class PathBuilder {
+ public:
+  explicit PathBuilder(bool start_nullable) : last_nullable_(start_nullable) {}
+  template <typename T>
+  void AddTerminalInfo(const T& array) {
+    if (last_nullable_) {
+      info_.max_def_level++;
+    }
+    if (array.null_count() == 0) {
```

Comment:
@wesm one point I wanted to ask about is the original writer.cc code used null_count() but it seems like it might be better to not compute this if it isn't already cached since we need to do the traversal anyways?  Was the choice there informed by benchmarks?

Reply:
I don't think it was informed by benchmarks